src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,343 @@
+            foreach (var signer in tx.Signers)
+            {
+                if (signer.Account == Hash)
+                {
+                    if (signer.Scopes != WitnessScope.None) return false;
+                    break;
+                }
+            }
+            if (tx.Sender == Hash)
+            {
+                if (tx.Signers.Length != 2) return false;
+                var payer = tx.Signers[1].Account;
+                var balance = GetDepositFor(engine.SnapshotCache, payer);
+                if (balance is null || balance.Amount.CompareTo(tx.NetworkFee + tx.SystemFee) < 0) return false;
+            }
+            var notaries = GetNotaryNodes(engine.SnapshotCache);
+            var hash = tx.GetSignData(engine.GetNetwork());
+            return notaries.Any(n => Crypto.VerifySignature(hash, signature, n));
+        }
+
+        /// <summary>
+        /// OnNEP17Payment is a callback that accepts GAS transfer as Notary deposit.
+        /// It also sets the deposit's lock height after which deposit can be withdrawn.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">GAS sender</param>
+        /// <param name="amount">The amount of GAS sent</param>
+        /// <param name="data">Deposit-related data: optional To value (treated as deposit owner if set) and Till height after which deposit can be withdrawn </param>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        private void OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash) throw new InvalidOperationException(string.Format("only GAS can be accepted for deposit, got {0}", engine.CallingScriptHash.ToString()));
+            if (data is not Array additionalParams || additionalParams.Count != 2) throw new FormatException("`data` parameter should be an array of 2 elements");
+            var to = from;
+            if (!additionalParams[0].Equals(StackItem.Null)) to = additionalParams[0].GetSpan().ToArray().AsSerializable<UInt160>();
+            var till = (uint)additionalParams[1].GetInteger();
+            var tx = (Transaction)engine.ScriptContainer;
+            var allowedChangeTill = tx.Sender == to;
+            var currentHeight = Ledger.CurrentIndex(engine.SnapshotCache);
+            if (till < currentHeight + 2) throw new ArgumentOutOfRangeException(string.Format("`till` shouldn't be less than the chain's height {0} + 1", currentHeight + 2));
+            // Don't need to seal because Deposit is a fixed-sized interoperable, hence always can be serialized.
+            var deposit = engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_Deposit, to))?.GetInteroperable<Deposit>();
+            if (deposit != null && till < deposit.Till) throw new ArgumentOutOfRangeException(string.Format("`till` shouldn't be less than the previous value {0}", deposit.Till));
```

Comment:
So, if an address makes a wrong till (too high) it can never be reversed?

Reply:
Yes, it's an intention.